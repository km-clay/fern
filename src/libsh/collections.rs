use std::collections::VecDeque;

pub trait VecDequeAliases<T> {
	fn fpop(&mut self) -> Option<T>;
	fn fpush(&mut self, value: T);
	fn bpop(&mut self) -> Option<T>;
	fn bpush(&mut self, value: T);
	fn to_vec(self) -> Vec<T>;
}

impl<T> VecDequeAliases<T> for VecDeque<T> {
	/// Alias for pop_front()
	fn fpop(&mut self) -> Option<T> {
		self.pop_front()
	}
	/// Alias for push_front()
	fn fpush(&mut self, value: T) {
		self.push_front(value);
	}
	/// Alias for pop_back()
	fn bpop(&mut self) -> Option<T> {
		self.pop_back()
	}
	/// Alias for push_back()
	fn bpush(&mut self, value: T) {
		self.push_back(value);
	}
	/// Just turns the deque into a vector
	fn to_vec(mut self) -> Vec<T> {
		let mut vec = vec![];
		while let Some(item) = self.fpop() {
			vec.push(item)
		}
		vec
	}
}
