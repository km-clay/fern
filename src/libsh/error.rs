use std::fmt::Display;

use crate::{
	libsh::term::{Style, Styled},
	parse::lex::Span,
	prelude::*
};

pub type ShResult<T> = Result<T,ShErr>;

pub trait ShResultExt {
	fn blame(self, span: Span) -> Self;
	fn try_blame(self, span: Span) -> Self;
}

impl<T> ShResultExt for Result<T,ShErr> {
	/// Blame a span for an error
	fn blame(self, new_span: Span) -> Self {
		let Err(e) = self else {
			return self
		};
		match e {
			ShErr::Simple { kind, msg, notes } |
			ShErr::Full { kind, msg, notes, span: _ } => Err(ShErr::Full { kind: kind.clone(), msg: msg.clone(), notes: notes.clone(), span: new_span }),
		}
	}
	/// Blame a span if no blame has been assigned yet
	fn try_blame(self, new_span: Span) -> Self {
		let Err(e) = &self else {
			return self
		};
		match e {
			ShErr::Simple { kind, msg, notes } => Err(ShErr::Full { kind: kind.clone(), msg: msg.clone(), notes: notes.clone(), span: new_span }),
			ShErr::Full { kind: _, msg: _, span: _, notes: _ } => self
		}
	}
}

#[derive(Clone,Debug)]
pub struct Note {
	main: String,
	sub_notes: Vec<Note>,
	depth: usize
}

impl Note {
	pub fn new(main: impl Into<String>) -> Self {
		Self { main: main.into(), sub_notes: vec![], depth: 0 }
	}

	pub fn with_sub_notes(self, new_sub_notes: Vec<impl Into<String>>) -> Self {
		let Self { main, mut sub_notes, depth } = self;
		for raw_note in new_sub_notes {
			let mut note = Note::new(raw_note);
			note.depth = self.depth + 1;
			sub_notes.push(note);
		}
		Self { main, sub_notes, depth }
	}
}

impl Display for Note {
	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
		let note = "note".styled(Style::Green);
		let main = &self.main;
		if self.depth == 0 {
			writeln!(f, "{note}: {main}")?;
		} else {
			let bar_break = "-".styled(Style::Cyan | Style::Bold);
			let indent = "  ".repeat(self.depth);
			writeln!(f, "  {indent}{bar_break} {main}")?;
		}

		for sub_note in &self.sub_notes {
			write!(f, "{sub_note}")?;
		}
		Ok(())
	}
}

#[derive(Debug)]
pub enum ShErr {
	Simple { kind: ShErrKind, msg: String, notes: Vec<Note> },
	Full { kind: ShErrKind, msg: String, notes: Vec<Note>, span: Span }
}

impl ShErr {
	pub fn simple(kind: ShErrKind, msg: impl Into<String>) -> Self {
		let msg = msg.into();
		Self::Simple { kind, msg, notes: vec![] }
	}
	pub fn full(kind: ShErrKind, msg: impl Into<String>, span: Span) -> Self {
		let msg = msg.into();
		Self::Full { kind, msg, span, notes: vec![] }
	}
	pub fn unpack(self) -> (ShErrKind,String,Vec<Note>,Option<Span>) {
		match self {
			ShErr::Simple { kind, msg, notes } => (kind,msg,notes,None),
			ShErr::Full { kind, msg, notes, span } => (kind,msg,notes,Some(span))
		}
	}
	pub fn with_note(self, note: Note) -> Self {
		let (kind,msg,mut notes,span) = self.unpack();
		notes.push(note);
		if let Some(span) = span {
			Self::Full { kind, msg, notes, span }
		} else {
			Self::Simple { kind, msg, notes }
		}
	}
	pub fn with_span(sherr: ShErr, span: Span) -> Self {
		let (kind,msg,notes,_) = sherr.unpack();
		let span = span.into();
		Self::Full { kind, msg, notes, span }
	}
	pub fn kind(&self) -> &ShErrKind {
		match self {
			ShErr::Simple { kind, msg: _, notes: _ } |
			ShErr::Full { kind, msg: _, notes: _, span: _ } => kind
		}
	}
	pub fn get_window(&self) -> Vec<(usize,String)> {
		let ShErr::Full { kind: _, msg: _, notes: _, span } = self else {
			unreachable!()
		};
		let mut total_len: usize = 0;
		let mut total_lines: usize = 1;
		let mut lines = vec![];
		let mut cur_line = String::new();

		let src = span.get_source();
		let mut chars = src.chars();

		while let Some(ch) = chars.next() {
			total_len += ch.len_utf8();
			cur_line.push(ch);
			if ch == '\n' {
				if total_len > span.start {
					let line = (
						total_lines,
						mem::take(&mut cur_line)
					);
					lines.push(line);
				}
				if total_len >= span.end {
					break
				}
				total_lines += 1;

				cur_line.clear();
			}
		}

		if !cur_line.is_empty() {
			let line = (
				total_lines,
				mem::take(&mut cur_line)
			);
			lines.push(line);
		}

		lines
	}
	pub fn get_line_col(&self) -> (usize,usize) {
		let ShErr::Full { kind: _, msg: _, notes: _, span } = self else {
			unreachable!()
		};

		let mut lineno = 1;
		let mut colno = 1;
		let src = span.get_source();
		let mut chars = src.chars().enumerate();
		while let Some((pos,ch)) = chars.next() {
			if pos >= span.start {
				break
			}
			if ch == '\n' {
				lineno += 1;
				colno = 1;
			} else {
				colno += 1;
			}
		}
		(lineno,colno)
	}
	pub fn get_indicator_lines(&self) -> Option<Vec<String>> {
		match self {
			ShErr::Simple { kind: _, msg: _, notes: _ } => None,
			ShErr::Full { kind: _, msg: _, notes: _, span } => {
				let text = span.as_str();
				let lines = text.lines();
				let mut indicator_lines = vec![];

				for line in lines {
					let indicator_line = "^".repeat(line.len()).styled(Style::Red | Style::Bold);
					indicator_lines.push(indicator_line);
				}

				Some(indicator_lines)
			}
		}
	}
}

impl Display for ShErr {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		match self {
			Self::Simple { msg, kind: _, notes } => {
				let mut all_strings = vec![msg.to_string()];
				let mut notes_fmt = vec![];
				for note in notes {
					let fmt = format!("{note}");
					notes_fmt.push(fmt);
				}
				all_strings.append(&mut notes_fmt);
				let mut output = all_strings.join("\n");
				output.push('\n');

				writeln!(f, "{}", output)
			}

			Self::Full { msg, kind, notes, span: _ } => {
				let window = self.get_window();
				let mut indicator_lines = self.get_indicator_lines().unwrap().into_iter();
				let mut lineno_pad_count = 0;
				for (lineno,_) in window.clone() {
					if lineno.to_string().len() > lineno_pad_count {
						lineno_pad_count = lineno.to_string().len() + 1
					}
				}
				let padding = " ".repeat(lineno_pad_count);
				writeln!(f)?;


				let (line,col) = self.get_line_col();
				let line_fmt = line.styled(Style::Cyan | Style::Bold);
				let col_fmt = col.styled(Style::Cyan | Style::Bold);
				let kind = kind.styled(Style::Red | Style::Bold);
				let arrow = "->".styled(Style::Cyan | Style::Bold);
				writeln!(f,
					"{kind} - {msg}",
				)?;
				writeln!(f,
					"{padding}{arrow} [{line_fmt};{col_fmt}]",
				)?;

				let mut bar = format!("{padding}|");
				bar = bar.styled(Style::Cyan | Style::Bold);
				writeln!(f,"{bar}")?;

				let mut first_ind_ln = true;
				for (lineno,line) in window {
					let lineno = lineno.to_string();
					let line = line.trim();
					let mut prefix = format!("{padding}|");
					prefix.replace_range(0..lineno.len(), &lineno);
					prefix = prefix.styled(Style::Cyan | Style::Bold);
					writeln!(f,"{prefix} {line}")?;

					if let Some(ind_ln) = indicator_lines.next() {
						if first_ind_ln {
							let ind_ln_padding = " ".repeat(col);
							let ind_ln = format!("{ind_ln_padding}{ind_ln}");
							writeln!(f, "{bar}{ind_ln}")?;
							first_ind_ln = false;
						} else {
							writeln!(f, "{bar} {ind_ln}")?;
						}
					}
				}

				write!(f,"{bar}")?;


				let bar_break = "-".styled(Style::Cyan | Style::Bold);
				if !notes.is_empty() {
					writeln!(f)?;
				}
				for note in notes {

					write!(f,
						"{padding}{bar_break} {note}"
					)?;
				}
				Ok(())
			}
		}
	}
}

impl From<std::io::Error> for ShErr {
	fn from(_: std::io::Error) -> Self {
		let msg = std::io::Error::last_os_error();
		ShErr::simple(ShErrKind::IoErr, msg.to_string())
	}
}

impl From<std::env::VarError> for ShErr {
	fn from(value: std::env::VarError) -> Self {
		ShErr::simple(ShErrKind::InternalErr, &value.to_string())
	}
}

impl From<rustyline::error::ReadlineError> for ShErr {
	fn from(value: rustyline::error::ReadlineError) -> Self {
		ShErr::simple(ShErrKind::ParseErr, &value.to_string())
	}
}

impl From<Errno> for ShErr {
	fn from(value: Errno) -> Self {
		ShErr::simple(ShErrKind::Errno, &value.to_string())
	}
}

#[derive(Debug,Clone)]
pub enum ShErrKind {
	IoErr,
	SyntaxErr,
	ParseErr,
	InternalErr,
	ExecFail,
	ResourceLimitExceeded,
	BadPermission,
	Errno,
	FileNotFound(String),
	CmdNotFound(String),
	CleanExit(i32),
	FuncReturn(i32),
	LoopContinue(i32),
	LoopBreak(i32),
	Null
}

impl Display for ShErrKind {
	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
		let output = match self {
			ShErrKind::IoErr => "I/O Error",
			ShErrKind::SyntaxErr => "Syntax Error",
			ShErrKind::ParseErr => "Parse Error",
			ShErrKind::InternalErr => "Internal Error",
			ShErrKind::ExecFail => "Execution Failed",
			ShErrKind::ResourceLimitExceeded => "Resource Limit Exceeded",
			ShErrKind::BadPermission => "Bad Permissions",
			ShErrKind::Errno => "ERRNO",
			ShErrKind::FileNotFound(file) => &format!("File not found: {file}"),
			ShErrKind::CmdNotFound(cmd) => &format!("Command not found: {cmd}"),
			ShErrKind::CleanExit(_) => "",
			ShErrKind::FuncReturn(_) => "",
			ShErrKind::LoopContinue(_) => "",
			ShErrKind::LoopBreak(_) => "",
			ShErrKind::Null => "",
		};
		write!(f,"{output}")
	}
}
