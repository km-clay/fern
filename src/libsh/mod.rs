pub mod error;
pub mod term;
