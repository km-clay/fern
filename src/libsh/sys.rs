use std::fmt::Display;

use crate::prelude::*;

pub const SIG_EXIT_OFFSET: i32 = 128;

pub fn get_bin_path(command: &str, shenv: &ShEnv) -> Option<PathBuf> {
	let env = shenv.vars().env();
	let path_var = env.get("PATH")?;
	let mut paths = path_var.split(':');
	while let Some(raw_path) = paths.next() {
		let mut path = PathBuf::from(raw_path);
		path.push(command);
		//TODO: handle this unwrap
		if path.exists() {
			return Some(path)
		}
	}
	None
}

pub fn write_out(text: impl Display) -> ShResult<()> {
	write(borrow_fd(1), text.to_string().as_bytes())?;
	Ok(())
}

pub fn write_err(text: impl Display) -> ShResult<()> {
	write(borrow_fd(2), text.to_string().as_bytes())?;
	Ok(())
}

/// Return is `readpipe`, `writepipe`
/// Contains all of the necessary boilerplate for grabbing two pipe fds using libc::pipe()
pub fn c_pipe() -> Result<(RawFd,RawFd),Errno> {
	let mut pipes: [i32;2] = [0;2];
	let ret = unsafe { libc::pipe(pipes.as_mut_ptr()) };
	if ret < 0 {
		return Err(Errno::from_raw(ret))
	}
	Ok((pipes[0],pipes[1]))
}

pub fn execvpe(cmd: String, argv: Vec<String>, envp: Vec<String>) -> Result<(),Errno> {
	let cmd_raw = CString::new(cmd).unwrap();

	let argv = argv.into_iter().map(|arg| CString::new(arg).unwrap()).collect::<Vec<CString>>();
	let envp = envp.into_iter().map(|var| CString::new(var).unwrap()).collect::<Vec<CString>>();

	nix::unistd::execvpe(&cmd_raw, &argv, &envp).unwrap();
	Ok(())
}
