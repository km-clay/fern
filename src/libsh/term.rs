use std::{fmt::Display, ops::BitOr};

pub trait Styled: Sized + Display {
	fn styled<S: Into<StyleSet>>(self, style: S) -> String {
		let styles: StyleSet = style.into();
		let reset = Style::Reset;
		format!("{styles}{self}{reset}")
	}
}

impl<T: Display> Styled for T {}

/// Enum representing a single ANSI style
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Style {
	// Undoes all styles
	Reset,
	// Foreground Colors
	Black,
	Red,
	Green,
	Yellow,
	Blue,
	Magenta,
	Cyan,
	White,
	BrightBlack,
	BrightRed,
	BrightGreen,
	BrightYellow,
	BrightBlue,
	BrightMagenta,
	BrightCyan,
	BrightWhite,
	RGB(u8, u8, u8),  // Custom foreground color

	// Background Colors
	BgBlack,
	BgRed,
	BgGreen,
	BgYellow,
	BgBlue,
	BgMagenta,
	BgCyan,
	BgWhite,
	BgBrightBlack,
	BgBrightRed,
	BgBrightGreen,
	BgBrightYellow,
	BgBrightBlue,
	BgBrightMagenta,
	BgBrightCyan,
	BgBrightWhite,
	BgRGB(u8, u8, u8),  // Custom background color

	// Text Attributes
	Bold,
	Dim,
	Italic,
	Underline,
	Strikethrough,
	Reversed,
}

impl Display for Style {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		match self {
			Style::Reset => write!(f, "\x1b[0m"),

			// Foreground colors
			Style::Black => write!(f, "\x1b[30m"),
			Style::Red => write!(f, "\x1b[31m"),
			Style::Green => write!(f, "\x1b[32m"),
			Style::Yellow => write!(f, "\x1b[33m"),
			Style::Blue => write!(f, "\x1b[34m"),
			Style::Magenta => write!(f, "\x1b[35m"),
			Style::Cyan => write!(f, "\x1b[36m"),
			Style::White => write!(f, "\x1b[37m"),
			Style::BrightBlack => write!(f, "\x1b[90m"),
			Style::BrightRed => write!(f, "\x1b[91m"),
			Style::BrightGreen => write!(f, "\x1b[92m"),
			Style::BrightYellow => write!(f, "\x1b[93m"),
			Style::BrightBlue => write!(f, "\x1b[94m"),
			Style::BrightMagenta => write!(f, "\x1b[95m"),
			Style::BrightCyan => write!(f, "\x1b[96m"),
			Style::BrightWhite => write!(f, "\x1b[97m"),
			Style::RGB(r, g, b) => write!(f, "\x1b[38;2;{r};{g};{b}m"),

			// Background colors
			Style::BgBlack => write!(f, "\x1b[40m"),
			Style::BgRed => write!(f, "\x1b[41m"),
			Style::BgGreen => write!(f, "\x1b[42m"),
			Style::BgYellow => write!(f, "\x1b[43m"),
			Style::BgBlue => write!(f, "\x1b[44m"),
			Style::BgMagenta => write!(f, "\x1b[45m"),
			Style::BgCyan => write!(f, "\x1b[46m"),
			Style::BgWhite => write!(f, "\x1b[47m"),
			Style::BgBrightBlack => write!(f, "\x1b[100m"),
			Style::BgBrightRed => write!(f, "\x1b[101m"),
			Style::BgBrightGreen => write!(f, "\x1b[102m"),
			Style::BgBrightYellow => write!(f, "\x1b[103m"),
			Style::BgBrightBlue => write!(f, "\x1b[104m"),
			Style::BgBrightMagenta => write!(f, "\x1b[105m"),
			Style::BgBrightCyan => write!(f, "\x1b[106m"),
			Style::BgBrightWhite => write!(f, "\x1b[107m"),
			Style::BgRGB(r, g, b) => write!(f, "\x1b[48;2;{r};{g};{b}m"),

			// Text attributes
			Style::Bold => write!(f, "\x1b[1m"),
			Style::Dim => write!(f, "\x1b[2m"),  // New
			Style::Italic => write!(f, "\x1b[3m"),
			Style::Underline => write!(f, "\x1b[4m"),
			Style::Strikethrough => write!(f, "\x1b[9m"),  // New
			Style::Reversed => write!(f, "\x1b[7m"),
		}
	}
}

/// Struct representing a **set** of styles
#[derive(Debug, Default, Clone)]
pub struct StyleSet {
	styles: Vec<Style>,
}

impl StyleSet {
	pub fn new() -> Self {
		Self { styles: vec![] }
	}

	pub fn add(mut self, style: Style) -> Self {
		if !self.styles.contains(&style) {
			self.styles.push(style);
		}
		self
	}
}

impl Display for StyleSet {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		for style in &self.styles {
			style.fmt(f)?
		}
		Ok(())
	}
}

/// Allow OR (`|`) operator to combine multiple `Style` values into a `StyleSet`
impl BitOr for Style {
	type Output = StyleSet;

	fn bitor(self, rhs: Self) -> Self::Output {
		StyleSet::new().add(self).add(rhs)
	}
}

/// Allow OR (`|`) operator to combine `StyleSet` with `Style`
impl BitOr<Style> for StyleSet {
	type Output = StyleSet;

	fn bitor(self, rhs: Style) -> Self::Output {
		self.add(rhs)
	}
}

impl From<Style> for StyleSet {
	fn from(style: Style) -> Self {
		StyleSet::new().add(style)
	}
}
