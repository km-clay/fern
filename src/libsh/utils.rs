use std::collections::VecDeque;

use crate::parse::lex::{Span, Tk};
use crate::parse::{Redir, RedirType};
use crate::prelude::*;

pub trait VecDequeExt<T> {
	fn to_vec(self) -> Vec<T>;
}

pub trait TkVecUtils<Tk> {
	fn get_span(&self) -> Option<Span>;
	fn debug_tokens(&self);
}

pub trait RedirVecUtils<Redir> {
	/// Splits the vector of redirections into two vectors
	///
	/// One vector contains input redirs, the other contains output redirs
	fn split_by_channel(self) -> (Vec<Redir>,Vec<Redir>);
}

impl<T> VecDequeExt<T> for VecDeque<T> {
	fn to_vec(self) -> Vec<T> {
		self.into_iter().collect::<Vec<T>>()
	}
}

impl TkVecUtils<Tk> for Vec<Tk> {
	fn get_span(&self) -> Option<Span> {
		if let Some(first_tk) = self.first() {
			if let Some(last_tk) = self.last() {
				Some(
					Span::new(
						first_tk.span.start..last_tk.span.end,
						first_tk.source()
					)
				)
			} else {
				None
			}
		} else {
			None
		}
	}
	fn debug_tokens(&self) {
		for token in self {
			flog!(DEBUG, "token: {}",token)
		}
	}
}

impl RedirVecUtils<Redir> for Vec<Redir> {
	fn split_by_channel(self) -> (Vec<Redir>,Vec<Redir>) {
		let mut input = vec![];
		let mut output = vec![];
		for redir in self {
			match redir.class {
				RedirType::Input => input.push(redir),
				RedirType::Pipe => {
					match redir.io_mode.tgt_fd() {
						STDIN_FILENO => input.push(redir),
						STDOUT_FILENO |
							STDERR_FILENO => output.push(redir),
						_ => unreachable!()
					}
				}
				_ => output.push(redir)
			}
		}
		(input,output)
	}
}
