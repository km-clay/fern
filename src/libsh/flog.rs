use std::fmt::Display;

use super::term::{Style, Styled};

#[derive(Clone, Copy, PartialEq, PartialOrd, Ord, Eq , Debug)]
#[repr(u8)]
pub enum FernLogLevel {
	NONE  = 0,
	ERROR = 1,
	WARN  = 2,
	INFO  = 3,
	DEBUG = 4,
	TRACE = 5
}

impl Display for FernLogLevel {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		use FernLogLevel::*;
		match self {
			ERROR => write!(f,"{}","ERROR".styled(Style::Red | Style::Bold)),
			WARN => write!(f,"{}","WARN".styled(Style::Yellow | Style::Bold)),
			INFO => write!(f,"{}","INFO".styled(Style::Green | Style::Bold)),
			DEBUG => write!(f,"{}","DEBUG".styled(Style::Magenta | Style::Bold)),
			TRACE => write!(f,"{}","TRACE".styled(Style::Blue | Style::Bold)),
			NONE => write!(f,"")
		}
	}
}

pub fn log_level() -> FernLogLevel {
	use FernLogLevel::*;
	let level = std::env::var("FERN_LOG_LEVEL").unwrap_or_default();
	match level.to_lowercase().as_str() {
		"error" => ERROR,
		"warn" =>  WARN,
		"info" =>  INFO,
		"debug" => DEBUG,
		"trace" => TRACE,
		_ =>       NONE
	}
}

/// A structured logging macro designed for `fern`.
///
/// `flog!` was implemented because `rustyline` uses `env_logger`, which clutters the debug output.
/// This macro prints log messages in a structured format, including the log level, filename, and line number.
///
/// # Usage
///
/// The macro supports three types of arguments:
///
/// ## 1. **Formatted Messages**
/// Similar to `println!` or `format!`, allows embedding values inside a formatted string.
///
/// ```rust
/// flog!(ERROR, "foo is {}", foo);
/// ```
/// **Output:**
/// ```plaintext
/// [ERROR][file.rs:10] foo is <value of foo>
/// ```
///
/// ## 2. **Literals**
/// Directly prints each literal argument as a separate line.
///
/// ```rust
/// flog!(WARN, "foo", "bar");
/// ```
/// **Output:**
/// ```plaintext
/// [WARN][file.rs:10] foo
/// [WARN][file.rs:10] bar
/// ```
///
/// ## 3. **Expressions**
/// Logs the evaluated result of each given expression, displaying both the expression and its value.
///
/// ```rust
/// flog!(INFO, 1.min(2));
/// ```
/// **Output:**
/// ```plaintext
/// [INFO][file.rs:10] 1
/// ```
///
/// # Considerations
/// - This macro uses `eprintln!()` internally, so its formatting rules must be followed.
/// - **Literals and formatted messages** require arguments that implement [`std::fmt::Display`].
/// - **Expressions** require arguments that implement [`std::fmt::Debug`].
#[macro_export]
macro_rules! flog {
	($level:path, $fmt:literal, $($args:expr),+ $(,)?) => {{
		use $crate::libsh::flog::log_level;
		use $crate::libsh::term::Styled;
		use $crate::libsh::term::Style;

		if $level <= log_level() {
			let file = file!().styled(Style::Cyan);
			let line = line!().to_string().styled(Style::Cyan);

			eprintln!(
				"[{}][{}:{}] {}",
				$level, file, line, format!($fmt, $($args),+)
			);
		}
	}};

	($level:path, $($val:expr),+ $(,)?) => {{
		use $crate::libsh::flog::log_level;
		use $crate::libsh::term::Styled;
		use $crate::libsh::term::Style;

		if $level <= log_level() {
			let file = file!().styled(Style::Cyan);
			let line = line!().to_string().styled(Style::Cyan);

			$(
				let val_name = stringify!($val);
				eprintln!(
					"[{}][{}:{}] {} = {:#?}",
					$level, file, line, val_name, &$val
				);
			)+
		}
	}};

	($level:path, $($lit:literal),+ $(,)?) => {{
		use $crate::libsh::flog::log_level;
		use $crate::libsh::term::Styled;
		use $crate::libsh::term::Style;

		if $level <= log_level() {
			let file = file!().styled(Style::Cyan);
			let line = line!().to_string().styled(Style::Cyan);

			$(
				eprintln!(
					"[{}][{}:{}] {}",
					$level, file, line, $lit
				);
			)+
		}
	}};
}
