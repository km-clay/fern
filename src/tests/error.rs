use libsh::error::{ShErr, ShErrKind};

use super::super::*;

#[test]
fn cmd_not_found() {
	let input = "foo";
	let token = LexStream::new(Rc::new(input.into()), LexFlags::empty()).next().unwrap().unwrap();
	let err = ShErr::full(ShErrKind::CmdNotFound("foo".into()), "", token.span);

	let err_fmt = format!("{err}");
	insta::assert_snapshot!(err_fmt)
}

#[test]
fn if_no_fi() {
	let input = "if foo; then bar;";
	let tokens = LexStream::new(Rc::new(input.into()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect::<Vec<_>>();

	let node = ParseStream::new(tokens).next().unwrap();
	let Err(e) = node else { panic!() };

	let err_fmt = format!("{e}");
	insta::assert_snapshot!(err_fmt)
}

#[test]
fn if_no_then() {
	let input = "if foo; bar; fi";
	let tokens = LexStream::new(Rc::new(input.into()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect::<Vec<_>>();

	let node = ParseStream::new(tokens).next().unwrap();
	let Err(e) = node else { panic!() };

	let err_fmt = format!("{e}");
	insta::assert_snapshot!(err_fmt)
}

#[test]
fn loop_no_done() {
	let input = "while true; do echo foo;";
	let tokens = LexStream::new(Rc::new(input.into()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect::<Vec<_>>();

	let node = ParseStream::new(tokens).next().unwrap();
	let Err(e) = node else { panic!() };

	let err_fmt = format!("{e}");
	insta::assert_snapshot!(err_fmt)
}

#[test]
fn loop_no_do() {
	let input = "while true; echo foo; done";
	let tokens = LexStream::new(Rc::new(input.into()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect::<Vec<_>>();

	let node = ParseStream::new(tokens).next().unwrap();
	let Err(e) = node else { panic!() };

	let err_fmt = format!("{e}");
	insta::assert_snapshot!(err_fmt)
}

#[test]
fn case_no_esac() {
	let input = "case foo in foo) bar;; bar) foo;;";
	let tokens = LexStream::new(Rc::new(input.into()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect::<Vec<_>>();

	let node = ParseStream::new(tokens).next().unwrap();
	let Err(e) = node else { panic!() };

	let err_fmt = format!("{e}");
	insta::assert_snapshot!(err_fmt)
}

#[test]
fn case_no_in() {
	let input = "case foo foo) bar;; bar) foo;; esac";
	let tokens = LexStream::new(Rc::new(input.into()), LexFlags::empty())
		.map(|tk| tk.unwrap())
		.collect::<Vec<_>>();

	let node = ParseStream::new(tokens).next().unwrap();
	let Err(e) = node else { panic!() };

	let err_fmt = format!("{e}");
	insta::assert_snapshot!(err_fmt)
}
