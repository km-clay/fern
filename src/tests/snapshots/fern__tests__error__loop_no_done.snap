---
source: src/tests/error.rs
expression: err_fmt
---
  [36m[1m->[0m [[36m[1m1[0m;[36m[1m1[0m] - [31m[1mParse Error[0m
[36m[1m  |[0m
[36m[1m1 |[0m while true; do echo foo;
[36m[1m  |[0m
  [36m[1m-[0m Expected 'done' after loop body
